-Wmismatched-tags in 3.5.0?

While I'm aware of the workarounds, I thought you might find the following interesting.

- What is the issue you have?

`-Wmismatched-tags` using Clang v6.0 on Ubuntu 18.04:

```
json.hpp:1799:1: error: 'tuple_size' defined as a class template here but previously declared as a struct template [-Werror,-Wmismatched-tags]
class tuple_size<::nlohmann::detail::iteration_proxy_value<IteratorType>>
^
/usr/include/c++/7.3.0/utility:84:5: note: did you mean class here?
    struct tuple_size;
    ^
[...]
json.hpp:1803:1: error: 'tuple_element' defined as a class template here but previously declared as a struct template [-Werror,-Wmismatched-tags]
class tuple_element<N, ::nlohmann::detail::iteration_proxy_value<IteratorType >>
^
/usr/include/c++/7.3.0/utility:129:5: note: did you mean class here?
    struct tuple_element;
    ^
2 errors generated.
```

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?

Attempted to update [Warzone 2100](https://github.com/Warzone2100/warzone2100) from nlohmann/json 3.4.0 to 3.5.0.
(Compilation succeeds with 3.4.0. CI treats many warnings as errors. `-std=gnu++11`)

- What is the expected behavior?

Compilation succeeds without error / warnings.

- And what is the actual behavior instead?

[-Wmismatched-tags] is triggered

- Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?

The CXX compiler identification is Clang 6.0.0
on Ubuntu 18.04

- Did you use a released version of the library or the version from the `develop` branch?

3.5.0 release

--- a/include/nlohmann/detail/iterators/iteration_proxy.hpp
+++ b/include/nlohmann/detail/iterators/iteration_proxy.hpp
@@ -147,6 +147,11 @@ auto get(const nlohmann::detail::iteration_proxy_value<IteratorType>& i) -> decl
 // And see https://github.com/nlohmann/json/pull/1391
 namespace std
 {
+#if defined(__clang__)
+    // Fix: https://github.com/nlohmann/json/issues/1401
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wmismatched-tags"
+#endif
 template <typename IteratorType>
 class tuple_size<::nlohmann::detail::iteration_proxy_value<IteratorType>>
             : public std::integral_constant<std::size_t, 2> {};
@@ -159,4 +164,7 @@ class tuple_element<N, ::nlohmann::detail::iteration_proxy_value<IteratorType >>
                      get<N>(std::declval <
                             ::nlohmann::detail::iteration_proxy_value<IteratorType >> ()));
 };
+#if defined(__clang__)
+    #pragma clang diagnostic pop
+#endif
 }

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -1808,6 +1808,11 @@ auto get(const nlohmann::detail::iteration_proxy_value<IteratorType>& i) -> decl
 // And see https://github.com/nlohmann/json/pull/1391
 namespace std
 {
+#if defined(__clang__)
+    // Fix: https://github.com/nlohmann/json/issues/1401
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wmismatched-tags"
+#endif
 template <typename IteratorType>
 class tuple_size<::nlohmann::detail::iteration_proxy_value<IteratorType>>
             : public std::integral_constant<std::size_t, 2> {};
@@ -1820,7 +1825,11 @@ class tuple_element<N, ::nlohmann::detail::iteration_proxy_value<IteratorType >>
                      get<N>(std::declval <
                             ::nlohmann::detail::iteration_proxy_value<IteratorType >> ()));
 };
+#if defined(__clang__)
+    #pragma clang diagnostic pop
+#endif
 }
+
 
 namespace nlohmann
 {

